synapse/app/homeserver.py

```diff
@@ -361,9 +357,43 @@ def setup(config_options):
     logger.info("Database prepared in %s.", config.database_config['name'])
 
     hs.setup()
-    hs.start_listening()
 
+    @defer.inlineCallbacks
     def start():
+        # Check if the certificate is still valid.
+        cert_days_remaining = hs.config.is_disk_cert_valid()
+
+        if hs.config.acme_enabled:
+            # If ACME is enabled, we might need to provision a certificate
+            # before starting.
+            acme = hs.get_acme_handler()
+
+            # Start up the webservices which we will respond to ACME challenges
+            # with.
+            try:
+                yield acme.start_listening()
+            except defer.FirstError as e:
+                # We couldn't listen on a port, for some reason. Print the
+                # traceback to stderr and bail.
+                e.subFailure.printTraceback(sys.stderr)
+                sys.exit(1)
+
+            # We want to reprovision if is_valid_cert is None (meaning no
+            # certificate exists), or the days remaining number it returns is
+            # less than our re-registration threshold.
+            if (cert_days_remaining is None) or (
+                not cert_days_remaining > hs.config.acme_reprovision_threshold
+            ):
+                yield acme.provision_certificate()
+
+        # Read the certificate from disk and build the context factories for
+        # TLS.
+        hs.config._read_certificate()
```

Comment:
please can this lose its `_` if it's going to be public

Reply:
sure